src/MinimalDelegation.sol

```diff
@@ -100,31 +109,24 @@ contract MinimalDelegation is IERC7821, IKeyManagement, ERC1271, EIP712 {
     }
 
     // Execute a batch of calls according to the mode and any optionally provided opData
-    function _execute(bytes32, Call[] memory, bytes memory) private pure {
-        // TODO: unpack anything required from opData
-        // verify signature from within opData
-        // if signature is valid, execute the calls
-        revert("Not implemented");
-    }
-
-    // We currently only support calls initiated by the contract itself which means there are no checks needed on the target contract.
-    // In the future, other keys can make calls according to their key permissions and those checks will need to be added.
-    function _execute(bytes32 mode, Call[] memory calls) private {
-        bool shouldRevert = mode.shouldRevert();
-        for (uint256 i = 0; i < calls.length; i++) {
-            (bool success, bytes memory output) = _execute(calls[i]);
-            // Reverts with the first call that is unsuccessful if the EXEC_TYPE is set to force a revert.
-            if (!success && shouldRevert) revert IERC7821.CallFailed(output);
+    function _execute(bytes32 mode, Call[] memory calls, bytes calldata opData) private {
+        // The caller must be address(this) if the mode is batchedCall or opData is empty
+        if (mode.isBatchedCall() || opData.length == 0) {
+            _authorizeCaller();
+            return _execute(mode, calls, bytes32(0));
```

Comment:
i feel like we shouldnt have to even call checkCanExecute in this code path, we might need to branch the logic further/maybe do 2 different internal calls based on the mode? its just wasteful cause we know already that it will return true for this case